fprotect: add sys init to protect image

osaether opened the pull request:
There is no reason not to protect the image partition.
All areas that should be written to are stored in separate
partitions (storage, settings etc.)

Reopening of https://github.com/nrfconnect/sdk-nrf/pull/3264

Ref: NCSDK-6933
Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-6269
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

oyvindronningstad reviewed (changes requested):
fprotect already has tests, but they are in /tests/drivers/fprotect :P. See if they already test what you wanted to test here.

You'd be more than welcome to move them to tests/lib/ instead, it definitely makes more sense :)

osaether commented:
> fprotect already has tests, but they are in /tests/drivers/fprotect :P. See if they already test what you wanted to test here.
> 
> You'd be more than welcome to move them to tests/lib/ instead, it definitely makes more sense :)

I think we need this test as well. I moved this test to tests/drivers/fprotect/app so we preserve the history of the other tests.
I also removed the calls to the depricated flash_write_protection_set.

hakonfam commented on lib/fprotect/Kconfig:
```diff
@@ -11,6 +11,8 @@ config FPROTECT
 	  Use hardware peripherals (BPROT, ACL, or SPU) to write-protect
 	  regions of flash until next reset.
 
+if FPROTECT
+
```
Remove this and the endif

hakonfam commented on lib/fprotect/Kconfig:
```diff
@@ -65,3 +67,24 @@ config FPROTECT_ENABLE_IS_PROTECTED
+
+config FPROTECT_APP
+	bool "Lock image partition using FPROTECT"
+	depends on PARTITION_MANAGER_ENABLED
```
```suggestion
	depends on PARTITION_MANAGER_ENABLED && FPROTECT
```

oyvindronningstad commented on tests/drivers/fprotect/app/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+    platform_allow: nrf9160dk_nrf9160 nrf52840dk_nrf52840 nrf52dk_nrf52832
+      nrf5340dk_nrf5340_cpuapp
+    tags: fprotect
+    harness: console
+    harness_config:
+      type: multi_line
+      ordered: true
+      regex:
+        - "E: ***** BUS FAULT *****"
+        - "E:   Precise data bus error"
```
```suggestion
```
See if it works without the harness config.

If not we should add more precise regex lines.

oyvindronningstad commented:
@andvib how can we add this new test to CI?

andvib commented:
@oyvindronningstad I see the other tests in tests/drivers/fprotect/ have the 'b0' tag - can that tag be added to these tests as well? With the 'b0' tag these tests will be picked up and run in the bootloader CI plan

carlescufi commented:
@osaether please rebase

coderbyheart commented on tests/drivers/fprotect/app/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
```
```suggestion
# Copyright (c) 2022 Nordic Semiconductor
```

coderbyheart commented on tests/drivers/fprotect/app/src/main.c:
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2022 Nordic Semiconductor ASA
```

coderbyheart commented on tests/drivers/fprotect/app/src/main.c:
```diff
@@ -0,0 +1,63 @@
+static void test_fatal(void)
+{
+	zassert_equal(expected_fatal, actual_fatal,
+			"The wrong number of fatal error has occurred (e:%d != a:%d).\n",
```
```suggestion
			"The wrong number of fatal errors has occurred (e:%d != a:%d).\n",
```
